pkg/api/api.go

```diff
@@ -354,12 +363,24 @@ func (a *API) RegisterStoreGateway(s *storegateway.StoreGateway) {
 	a.RegisterRoute("/store-gateway/tenant/{tenant}/blocks", http.HandlerFunc(s.BlocksHandler), false, true, "GET")
 }
 
-// RegisterCompactor registers the ring UI page associated with the compactor.
+// RegisterCompactor registers routes associated with the compactor.
 func (a *API) RegisterCompactor(c *compactor.MultitenantCompactor) {
 	a.indexPage.AddLinks(defaultWeight, "Compactor", []IndexPageLink{
 		{Desc: "Ring status", Path: "/compactor/ring"},
 	})
 	a.RegisterRoute("/compactor/ring", http.HandlerFunc(c.RingHandler), false, true, "GET", "POST")
+	// Endpoint to handle requests for completion of block backfilling.
+	// Placed before less specific handler for same route, that doesn't take query parameter, for precedence.
+	a.RegisterRouteWithQueryParameters("/api/v1/upload/block/{block}", http.HandlerFunc(c.CompleteBlockUpload),
+		true, false, []string{"uploadComplete", "true"}, http.MethodPost)
+	// Endpoint to handle requests for starting of block backfilling.
+	// Starting the backfilling of a block means to verify that the backfill can go ahead.
+	// In practice this means to check that the block isn't already in block storage.
+	a.RegisterRoute("/api/v1/upload/block/{block}", http.HandlerFunc(c.CreateBlockUpload), true,
+		false, http.MethodPost)
+	// Endpoint to handle requests for uploading block files to a backfill.
+	a.RegisterRoute("/api/v1/upload/block/{block}/{path}", http.HandlerFunc(c.UploadBlockFile),
```

Comment:
Nit: I'd pass the path as a query param here instead of the url path, it needs to be urlencoded anyway, so having it as path isn't as pretty as it looks like:
`/api/v1/upload/block/FOOBAR/chunks%2Fchunk0001`.

That way we won't need to explicitly urldecode it in the handler.

Reply:
:thinking:

Reply:
@pstibrany can you help me understand how it concretely causes confusion/bugs? Is it because clients fail to realize they need to encode paths before placing them in URLs?

Reply:
Example of such bug: https://github.com/grafana/grafana/issues/42947 (interaction between Grafana and Cortex/Mimir's ruler)

Problem with requiring escaped paths like that is that we rely on all proxies between client and Mimir to not mangle it in any way. (And on client making it correct in the first place).

Reply:
Note the conclusion from that bug: 

> This was discussed further with the engineering team – consensus is not to try and fix this in the current API of Cortex / Loki / Grafana and to make sure we configure all proxies properly instead.
> We are instead spending our effort on an API redesign where this issue will be resolved.

We can start with better API from the beginning :)

Reply:
Thanks for providing context @pstibrany, that's very helpful!

Reply:
@pstibrany I pushed a revision sticking block file paths in a query parameter instead. Haven't tested it yet.